scripts/prepare-test-ledger.sh

```diff
@@ -0,0 +1,104 @@
+#!/usr/bin/env bash
+
+# Usage: ./scripts/prepare-test-ledger.sh <BP key 1> <BP key 2> ... <BP key n>
+
+# Number of keys in ledger that won't be re-delegated
+NUM_GHOST_KEYS=${NUM_GHOST_KEYS:-0}
+KEY_BALANCE=100000
+
+echo "Script assumes mainnet's start was at epoch 0 on 17th March 2021, if it's not the case, update the script please" >&2
+
+if [[ $# -eq 0 ]]; then
+  echo "No arguments specified: need block producing keys" >&2
+  exit 1
+fi
+
+num_keys=$#
+num_keys_and_ghost=$((num_keys + NUM_GHOST_KEYS))
+
+now=`date +%s`
+mainnet_start=`date --date='2021-03-17 00:00:00' -u +%s`
+epoch_now=$(( (now-mainnet_start)/7140/180 ))
+
+ledger_file="$epoch_now.json"
+
+echo "Current epoch: $epoch_now" >&2
+
+if [[ ! -f "$ledger_file" ]]; then
+  ledgers_url="https://storage.googleapis.com/storage/v1/b/mina-staking-ledgers/o?maxResults=1&prefix=staking-$epoch_now"
+  ledger_url=$(curl "$ledgers_url" | jq -r .items[0].mediaLink)
+
+  wget -O $ledger_file "$ledger_url"
+fi
+
+ixs_=$(for key in $@; do
+  <$ledger_file jq "[.[].pk] |index(\"$key\")"
+done)
+
+ixs=()
+i=0; for ix in $ixs_; do
```

Comment:
GENERAL QUESTION:

Looks like the script has some not trivial logic why bash was chosen as language?

Reply:
I thought about it, but I really like how the actual processing is done with `jq` command

I'm not on good terms with Python, honestly. And for other languages there are downsides of necessity to compile/ensure dependencies are installed/less expressive implementation because of absence of abstraction like `.jq`.

And initial implementation was a bit simpler, without the need to account for possibility of existence of the block producer keys in the ledger, this block that you're pointing to and some more would go away.

I may think of how to simplify this code though.

Reply:
ChatGPT is sometimes good at translation into python (overrepresented in their training), and plumbum https://plumbum.readthedocs.io/en/latest/ takes care of a lot of pain.

Reply:
I'd prefer to write it in Haskell, but this would make it more painful to use it (need to compile Haskell) and support it (would require someone speaking Haskell) :D

Reply:
Anyway, I will push another version of this loop later to simplify things. let's make the script under 100 lines